feat(optimism): Add support for all Superchain configs and enable their usage

cakevm opened the pull request:
Add functions to read Genesis JSON from a zip file, allowing large JSON files to be removed from the repository.

Requires to work: https://github.com/alloy-rs/alloy/pull/2073 to read the ChainConfig from the toml files

Added BSL-1.0 license. [See here for compatibility to MIT/Apache 2.0](https://wiki.geant.org/display/GSD/Important+licences+for+licence+selection#:~:text=The%20BSL%201.0%20licence%20is,licensed%20under%20Apache%20License%202.0.)

@mattsse: Seeking for advice how to get rid of `std:io`

mattsse reviewed (commented):
@clabby @refcell 

is there a use case for these static chainspecs in kona?

I guess it would depend how we proceed with #14571

we likely need to separate parts of the chainspec logic, like hardfork activations and maybe fee params not necessarily genesis

so we could feature gate these to std for now probably

clabby commented:
> @clabby @refcell
> 
> is there a use case for these static chainspecs in kona?
> 
> I guess it would depend how we proceed with #14571
> 
> we likely need to separate parts of the chainspec logic, like hardfork activations and maybe fee params not necessarily genesis
> 
> so we could feature gate these to std for now probably

Yeah, we could use them :D 

fwiw, @refcell already has some rust bindings set up for the [`superchain-registry`](https://github.com/ethereum-optimism/superchain-registry) [here](https://github.com/op-rs/maili/tree/main/crates/registry). We use these bindings to pull in our [`RollupConfig`](https://docs.rs/maili-genesis/latest/maili_genesis/struct.RollupConfig.html)s (think CL `ChainSpec`) from the upstream source of truth. It'd be awesome for us to saturate these `ChainSpec`s from that information. The genesis files over there are also [all compressed w/ `zstd`](https://github.com/ethereum-optimism/superchain-registry/tree/main/superchain/extra/genesis/mainnet).

cakevm commented:
@mattsse This PR is ready for review. `no-std` is finally working. Here some remarks:

- This PR is a step forward to read config + genesis from the Superchain registry files. I only did it now for `base` would like to keep this PR lean and get more generic with another PR.
- I could not find a reasonable way to handle zip, toml and zstd in `no-std`. For that reason the config files are preprocessed and converted to json. And the compression is changed.
- I added the module `superchain_registry` with the intention to extract this maybe later to alloy for others to use. To see what is required.
- I keept the "config:null" workaround until I got feedback from: https://github.com/ethereum-optimism/superchain-registry/issues/901

cakevm commented:
**Small update:** Now setting fields (e.g., `eip150_block: Some(0),`) like `op-geth` instead of defaulting to zero. I feel this makes comparisons easier. I skipped it before because I could not see the requirement.

If you serialize and compare `genesis` used in `BASE_MAINNET` and `BASE_SEPOLIA`, you will see new fields introduced by this PR:

base-mainnet:
```diff
14442a14443,14444
>     "cancunTime": 1710374401,
>     "canyonTime": 1704992401,
14445a14448,14449
>     "deltaTime": 1708560000,
>     "ecotoneTime": 1710374401,
14448a14453,14454
>     "fjordTime": 1720627201,
>     "graniteTime": 1726070401,
14449a14456
>     "holoceneTime": 1736445601,
14456a14464
>       "eip1559DenominatorCanyon": 250,
14460a14469
>     "shanghaiTime": 1704992401,
```

base-sepolia:
```diff
15224a15225,15226
>     "cancunTime": 1708534800,
>     "canyonTime": 1699981200,
15227a15230,15231
>     "deltaTime": 1703203200,
>     "ecotoneTime": 1708534800,
15230a15235,15236
>     "fjordTime": 1716998400,
>     "graniteTime": 1723478400,
15231a15238
>     "holoceneTime": 1732633200,
15238a15246
>       "eip1559DenominatorCanyon": 250,
15242a15251
>     "shanghaiTime": 1699981200,
```

cakevm commented:
Just a quick remark, I tried to do a full fresh sync for `unichain-mainnet` and it is failing currently.

I am still investigating what is wrong here. I was able to sync to head before, and can not tell right now what has changed.
```
2025-03-04T11:38:30.239429Z ERROR Stage encountered an execution error: EVM reported invalid transaction (0x6b37cec971c7f9c17d2355d737ffc0531b540b08d818d9a7c74bbf87098e7f7c): deposit transaction halted post-regolith; error will be bubbled up to main return handler stage=Execution bad_block=760073
2025-03-04T11:38:30.239514Z  INFO Unwinding{stage=Finish}: Starting unwind from=500001 to=500001 bad_block=Some(760073)
```

mattsse commented on crates/primitives-traits/src/transaction/recover.rs:
whoops,

do you want to submit this separately?

cakevm replied:
Yep, https://github.com/paradigmxyz/reth/pull/14861

mattsse commented on crates/optimism/chainspec/src/lib.rs:
can we add some docs here that cover the registry -> chainspec process?

cakevm replied:
I added some details about it, but as comment and not rust doc. Thought it is more an internal detail

mattsse commented on crates/optimism/chainspec/src/chain_specs.rs:
I think being explicit about chains that we support is good but I'd like to feature gate the more exit ones because we only need this for the cli, and keep base+optimism variants by default and the rest behind a feature

mattsse replied:
or maybe, because these are lazy, we don't actually need to do this?

cakevm replied:
I share this perspective. I think because they are lazy it doesn't matter

cakevm commented:
I rebased and have now the patch from https://github.com/paradigmxyz/reth/pull/14855. My unichain node is now able again to sync to HEAD.

mattsse commented on crates/optimism/chainspec/src/chain_metadata.rs:
this looks a lot like the registry support in kona

I think for reth this is out of scope and we can keep them to a minimum to what we actually need from registry files to derive the chainspec

cakevm replied:
I removed all not required fields and removed the dead_code exception

mattsse reviewed (changes requested):
all of this looks very good,

only have a few nits and suggestions

could you also show the diff in op-reth binary size main vs this

mattsse commented:
> Saving now about 10MB.

great, considering that this now has a ton of additional chainspecs

mattsse commented:
@emhane could perhaps also taking a look at this?

emhane commented on crates/optimism/chainspec/src/chain_metadata.rs:
```diff
@@ -0,0 +1,312 @@
+    pub optimism: Option<OptimismConfig>,
+}
+
+#[derive(Clone, Debug, Deserialize)]
```
this type already lives in kona, how do we want to dedup? @refcell @clabby @mattsse 
https://github.com/op-rs/kona/blob/a1899da1cfc0932b2f42fb799abdc5b558c1ba39/crates/protocol/genesis/src/chain/hardfork.rs#L3-L58

mattsse replied:
yeah wen't need these types really, and would get away with minal bindings so that we get access to what we need to put into the genesis object

emhane commented on crates/optimism/chainspec/src/chain_metadata.rs:
```diff
@@ -0,0 +1,312 @@
+    }
+}
+
+/// Returns a [`ChainConfig`] filled from [`ChainMetadata`] with extra fields and handling
```
same here, we probably want to use this type right? https://github.com/op-rs/kona/blob/a1899da1cfc0932b2f42fb799abdc5b558c1ba39/crates/protocol/genesis/src/chain/config.rs#L13-L106

emhane reviewed (commented):
some of this logic already lives in kona, seems inefficient to double this up due to maintenance

lemme know how we want to solve this @refcell @clabby @mattsse

cakevm commented:
@ConjunctiveNormalForm Feel free to try this branch if you like. I did not tried `unichain-sepolia` so far. Would be happy about any feedback. The mentioned fix is already included in this branch here.

cakevm commented:
I just want to make a remark here that I am not able to run `unichain-mainnet` for an extended period of time. However, this seems to be unrelated to this PR. Syncing to the head is possible, but after a short period, `op-node` stops sending RPC requests to `reth`, and `reth` gets stuck in a loop trying to unwind the state. It looks something like this:

```
2025-03-07T10:10:47.205079Z  INFO Received headers total=10000 from_block=10553885 to_block=10543886
2025-03-07T10:10:47.207710Z ERROR Cannot attach header to head error=mismatched parent hash: got 0x727333f68423b6d81e7bfe2e794b3205442ca2a0e3a2b77be5a5fe3f43d7bd60, expected 0x2e41f0ba3d0f05b1595281f0720a93b850d94275dc6708fce4ea8520be4064bd
2025-03-07T10:10:47.207723Z  WARN Stage encountered detached head stage=Headers local_head=BlockWithParent { parent: 0xcb910f7e12c3039a90a14e46539fe378338927243045eaed327bca1a9c6aac6c, block: NumHash { number: 10533105, hash: 0x2e41f0ba3d0f05b1595281f0720a93b850d94275dc6708fce4ea8520be4064bd } } header=BlockWithParent { parent: 0x727333f68423b6d81e7bfe2e794b3205442ca2a0e3a2b77be5a5fe3f43d7bd60, block: NumHash { number: 10533106, hash: 0x8af357c01ff6a83291c919160787be0540a60b8782c8fc467dd3c174c2c4ae34 } } error=mismatched parent hash: got 0x727333f68423b6d81e7bfe2e794b3205442ca2a0e3a2b77be5a5fe3f43d7bd60, expected 0x2e41f0ba3d0f05b1595281f0720a93b850d94275dc6708fce4ea8520be4064bd
2025-03-07T10:10:47.207769Z  INFO Unwinding{stage=Finish}: Starting unwind from=10533081 to=10533081 bad_block=Some(10533105)
2025-03-07T10:10:47.207780Z  INFO Unwinding{stage=IndexAccountHistory}: Starting unwind from=10533081 to=10533081 bad_block=Some(10533105)
2025-03-07T10:10:47.207786Z  INFO Unwinding{stage=IndexStorageHistory}: Starting unwind from=10533081 to=10533081 bad_block=Some(10533105)
2025-03-07T10:10:47.207790Z  INFO Unwinding{stage=TransactionLookup}: Starting unwind from=10533081 to=10533081 bad_block=Some(10533105)
2025-03-07T10:10:47.207795Z  INFO Unwinding{stage=MerkleExecute}: Starting unwind from=10533081 to=10533081 bad_block=Some(10533105)
2025-03-07T10:10:47.207799Z  INFO Unwinding{stage=StorageHashing}: Starting unwind from=10533081 to=10533081 bad_block=Some(10533105)
2025-03-07T10:10:47.207803Z  INFO Unwinding{stage=AccountHashing}: Starting unwind from=10533081 to=10533081 bad_block=Some(10533105)
2025-03-07T10:10:47.207807Z  INFO Unwinding{stage=MerkleUnwind}: Starting unwind from=10533081 to=10533081 bad_block=Some(10533105)
2025-03-07T10:10:47.207812Z  INFO Unwinding{stage=Execution}: Starting unwind from=10533081 to=10533081 bad_block=Some(10533105)
2025-03-07T10:10:47.207816Z  INFO Unwinding{stage=SenderRecovery}: Starting unwind from=10533081 to=10533081 bad_block=Some(10533105)
2025-03-07T10:10:47.207820Z  INFO Unwinding{stage=Bodies}: Starting unwind from=10533081 to=10533081 bad_block=Some(10533105)
2025-03-07T10:10:47.207824Z  INFO Unwinding{stage=Headers}: Starting unwind from=10533081 to=10533081 bad_block=Some(10533105)
2025-03-07T10:10:47.207948Z  WARN Bad block detected in unwind invalid_block=BlockWithParent { parent: 0xcb910f7e12c3039a90a14e46539fe378338927243045eaed327bca1a9c6aac6c, block: NumHash { number: 10533105, hash: 0x2e41f0ba3d0f05b1595281f0720a93b850d94275dc6708fce4ea8520be4064bd } }
2025-03-07T10:10:47.922001Z  INFO Preparing stage pipeline_stages=1/12 stage=Headers checkpoint=10533081 target=None
2025-03-07T10:10:47.922021Z  INFO Received headers total=10000 from_block=10593887 to_block=10583888
2025-03-07T10:10:47.931562Z  INFO Received headers total=10000 from_block=10583887 to_block=10573888
2025-03-07T10:10:47.939124Z  INFO Received headers total=10000 from_block=10573887 to_block=10563888
```

I have the feeling it is caused by a reorg that is happening, and `op-node` and `reth` are playing it not well. For `base-mainnet` everything is fine. I tired to increase `BEACON_CONSENSUS_REORG_UNWIND_DEPTH` to 24 (12*1sec block * 2 blocks). But ran then even in a total different issue. If someone is playing around with that feel free to reach out. But right now I move on to different topics.

klkvr reviewed (approved):
lgtm

cakevm commented:
Updated the superchains, and added the Isthmus activation for sepolia time to the test. See: https://github.com/ethereum-optimism/superchain-registry/pull/956

cakevm commented:
Rebase because it was now conflicting with https://github.com/paradigmxyz/reth/pull/15685

cakevm commented:
@mattsse Two tests for the latest fork id are now failing, with the latest update from superchain-registry that contain the Isthmus activation time.  (See: https://github.com/ethereum-optimism/superchain-registry/pull/972) 
It seems to be a easy fix to update the fork id, but since the `NetworkConfigBuilder` is using `latest_fork_id()` maybe this could cause issues in the consensus? I wonder if the `ForkCondition` should not always be evaluated?

emhane commented:
making a pr for adding mainnet timestamps from https://github.com/ethereum-optimism/superchain-registry/pull/972

cakevm commented:
> @mattsse Two tests for the latest fork id are now failing, with the latest update from superchain-registry that contain the Isthmus activation time. (See: [ethereum-optimism/superchain-registry#972](https://github.com/ethereum-optimism/superchain-registry/pull/972)) It seems to be a easy fix to update the fork id, but since the `NetworkConfigBuilder` is using `latest_fork_id()` maybe this could cause issues in the consensus? I wonder if the `ForkCondition` should not always be evaluated?

Thanks for fixing this with https://github.com/paradigmxyz/reth/pull/15769. I rebased again and updated to the latest configs.

mattsse commented on crates/chainspec/res/genesis/base.json:
I would like to keep those because this is a lot easier then going to the superchain repo and extracting them.

we can just keep them

cakevm replied:
I would argue that the test coverage for `base` will help to keep this feature stable. This was also my intention for removing the case for `base`. I would not mind to remove it, but then there are to cases to handle for the future. How do you feel about that?

mattsse commented on crates/net/peers/src/lib.rs:
unrelated please undo

mattsse commented on crates/optimism/chainspec/src/base.rs:
for this PR I'd like to keep all of the existing chainspecs that we introduced manually

it's easier to get this in that way

mattsse commented on crates/optimism/chainspec/src/base_sepolia.rs:
same

mattsse commented on crates/optimism/chainspec/src/chain_metadata.rs:
do we really need all of this, we already have a way to parse from genesis, so unclear why we need this

cakevm replied:
Not all genesis files are providing this data. We had this question already before and I created this issue here: https://github.com/ethereum-optimism/superchain-registry/issues/901 And there is additional logic from op-geth for reading it.

The motivation for the superchain-registry is probably to keep the footprint (adjusting files in git) low, for that reason there is a split of genesis and metadata for the superchain.

mattsse replied:
ah right, I forgot about this...

great 👍

mattsse commented on crates/optimism/chainspec/src/chain_specs.rs:
this seems fine but I'd like to feature gate this and only enable for cli because this is just bloat

mattsse reviewed (changes requested):
okay, I finally picked this up again.

I put this off for a while because dealing with chainspecs gives me headaches, especially if this changes the way we do this, so this change is critical.

therefor I'd like to make a few changes to this
* keep all `ChainSpec` static that we currently have, this way we can ensure that we don't accidentally break anything.

ideally we can also simplify the parsing and get rid of the additional types.

and we need some docs in the readme about how this all works.

cakevm commented on crates/optimism/chainspec/src/lib.rs:
```diff
@@ -9,15 +9,54 @@
 #![cfg_attr(not(test), warn(unused_crate_dependencies))]
 #![cfg_attr(not(feature = "std"), no_std)]
 
+// About the provided chain specs from `res/superchain-configs.tar`:
```
Find here some doc how it works

cakevm commented:
@mattsse Ready to review it again

cakevm commented:
@mattsse Ready to review it again (lint is failing caused by the unrelated const fn stuff)

mattsse reviewed (approved):
very nice, sorry that this took so long.

I have a few ideas for followups here but for the initial integration this is great
